File: 규진/프로그래머스/week16/연속_펄스_부분_수열의_합.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;
vector<int> v1;
vector<int> v2;
long long dp[500001];

long long getMaxSum(vector<int> &v) {
    long long ret = v[0];
    dp[0] = v[0];
    
    for (int i = 1; i < v.size(); i++) {
        dp[i] = max(dp[i - 1] , (long long) 0) + v[i];
        ret = max(ret, dp[i]);
    }
    return ret;
}

long long solution(vector<int> sequence) {
    long long answer = 0;
    for (int i = 0; i < sequence.size(); i++) {
        if (i % 2 == 0) {
            v1.push_back(sequence[i]);
            v2.push_back(-sequence[i]);
        } else {
            v1.push_back(-sequence[i]);
            v2.push_back(sequence[i]);
        }
    }
    long long r1 = getMaxSum(v1);
    fill(dp, dp + 500001, 0);
    long long r2 = getMaxSum(v2);
    
    answer = max(r1, r2);
    
    return answer;
}

File: 명진/programmers/연속 펄스 부분 수열의 합.py
def solution(sequence):
    n = len(sequence)
    dp = [[sequence[i] * (1 if i % 2 else -1) for i in range(n)]]
    dp += [[sequence[i] * (-1 if i % 2 else 1) for i in range(n)]]

    answer = max(sequence[0], -sequence[0])
    for i in range(2):
        for j in range(1, n):
            dp[i][j] += max(0, dp[i][j-1])
            answer = max(answer, dp[i][j])

    return answer